<?xml version="1.0" ?>
<!-- [25-9-24] Constructed for simulating a 3 DoF SMM, with 3 joint effort controllers && idosc centralized node -->
<!--           - idosc_master node executes the "update" function and publishes the joint torques 
               - the controller yaml is located in the ros_pkg of the robot xacro file: smm_ros_gazebo_pkg
-->
     
<launch>
    <!-- Include launch file Gazebo with robot and all controller settings -->
     <include file="$(find smm_control)/launch/idosc_gazebo_settings.launch"/>
    
    <!-- IDOSC SIMPLE IMPLEMENTATION -->
    
    <!-- Node that generates the topic /joint_current_state -->
    <node name="updateJointState_idosc_simple" pkg="smm_control" type="updateJointState_idosc_simple" output="screen" />

        <!-- Analytical Kinematics/Jacobians using screws @ /joint_current_state -->
        <node name="serverToolJacobians" pkg="smm_control" type="serverToolJacobians" output="screen" />

        <!-- Analytical Dynamics using screws @ /joint_current_state -->
        <node name="serverRobotDynamics" pkg="smm_control" type="serverRobotDynamics" output="screen" /> -->

    <!-- Node that generates the topic /tcp_current_state -->
    <node name="updateCartesianState_idosc_simple" pkg="smm_control" type="updateCartesianState_idosc_simple" output="screen" />

    <!-- Node that publishes to /tcp_desired_state -->
    <rosparam file="$(find smm_control)/config/yaml/paths/path_reid16.yaml" command="load"/>
    <node name="publishCartesianPathLoop_idosc_simple" pkg="smm_control" type="publishCartesianPathLoop_idosc_simple" output="screen" />

    <!-- Node that publishes to /idosc_error_state -->
    <node name="updateErrorState_idosc_simple" pkg="smm_control" type="updateErrorState_idosc_simple" output="screen" />

    <!-- Launch node for controller gains setting -->
    <node name="updateGains_idosc_simple" pkg="smm_control" type="updateGains_idosc_simple" output="screen" />

        <!-- Manual launch: rqt_reconfigure for manual gain setting -->
        <!--<node name="rqt_reconfigure" pkg="rqt_reconfigure" type="rqt_reconfigure" output="screen" />-->
    
    <!-- Launch controller node-->
    <node name="updateControlOutput_idosc_simple" pkg="smm_control" type="updateControlOutput_idosc_simple" output="screen" >
        <param name="rosconsole_config_file" value="$(find smm_control)/config/rosconsole.config"/>
    </node>

    <!-- Launch torque publisher node-->
    <node name="updateTorques_idosc_simple" pkg="smm_control" type="updateTorques_idosc_simple" output="screen" >
        <param name="rosconsole_config_file" value="$(find smm_control)/config/rosconsole.config"/>
    </node>

    <!-- Robot joints effort publisher node-->
    <node name="robot_effort_publisher" pkg="smm_control" type="robot_effort_publisher" output="screen" />

    <!-- Include launch file for data logging and real time monitoring -->
     <include file="$(find smm_control)/launch/idosc_data_monitor.launch"/>
</launch>